content/architecture/storage-engine/prolly-tree.md

```diff
@@ -2,38 +2,234 @@
+
+4. **Finished?**: If the length of your list of content addresses is 1, then you are done. This is the content address of your tree.
+
+5. **Build a New Map** Otherwise, form the next layer of your Prolly Tree by creating a map of the highest key value in the chunk and the content address of the chunk. Here's what the entries for the first internal level of the tree would look like for our example:
+
+![Building a Prolly-tree, Step 5](../../.gitbook/assets/tim-prolly-tree-step-5.png)
+
+6. **Return to Step 2**: Use this new map step 2. The algorithm terminates when step 4's condition is reached.
+
+Following these steps results in the following Prolly Tree. 
+
+![Prolly Tree](../../.gitbook/assets/tim-prolly-tree-example.png)
+
+## Modifying a Prolly Tree
+
+The magic of Prolly Trees is not seen on construction, but on modification. As you'll see, modifying a Prolly Tree changes the minimal amount of content addresses. Minimal modification means more structural sharing of contents because fewer content addresses change. It is also the basis of fast diff which is accomplished by comparing content addresses.
+
+### Update a Value
+
+If we update a value, we walk the tree by key to the leaf node holding the value. We then edit the referenced chunk in place. After the edit, we recalculate the content address of the chunk. We then walk up the tree recalculating each internal content address up to the root of the tree. 
+
+![Prolly Tree Value Update](../../.gitbook/assets/prolly-tree-update-value.png)
+
+### Insert Keys
+
+We can insert keys at the beginning, middle, or end of the key space. It's helpful to visualize each type of insert.
+
+When we insert a key, we walk the tree by key to the leaf node where the key belongs. We then edit the chunk, calculate whether to split the chunk or not, and then recalculate the content address. Note, a key insert or delete has a small probability of splitting the chunk. If the rolling hash value when incorporating the new key randomly comes up below our boundary threshold, the chunk is split at the new boundary. Finally, we walk up the tree recalculating each content address up to the root of the tree.
+
+In Dolt's Prolly Tree implementation, chunks are set to be on average 4 kilobytes. This means that you have a base probability of 1/4096 or 0.02% of triggering a chunk boundary shift with a single byte rolling hash window. A single byte window would not be enough data to provide the requisite level of randomness so the rolling hash window is set to 64 bytes, providing a base probability of a chunk split of 64/4096 or 1.6%. As we'll discuss later, the size of the chunk is considered when computing the rolling hash. So, the larger the chunk, the greater the probability the chunk will be split on the addition of a new key. Conversely, for small chunks, the probability of a chunk split is very small. This means that every edit to a table in Dolt is a minimum of 4Kb multiplied by the depth of the tree.
```

Comment:
This is a little muddied because we no longer use a rolling hash.

In reality, inserting or deleting a key has some probability of changing the existing chunking because of the following factors:

* The new key might happen to have a very "good" hash, which really wants to split.
* The removed key might happen to have had a very "good" hash, which really wanted to split at it.
* Adding or removing data from the chunk will change the "target value" all other keys are looking for when they compare their hashes on the re-chunking phase.

In reality, what happens is, the existing chunk picked a key because it was both
* already big enough
* the key had a pretty "good" hash value.

In particular, the keys which came directly before the key where the existing split is – their hashes are not dramatically better than its hash and many of them are likely to be worse.

Because we chunk on a per-key bases, the probability of splitting or re-chunking is more based on the size of the data mutation you make and on the number of keys that are in the chunk in general. For example, if you tuples that are ~128 bytes then there will only be ~32 tuples per chunk, and then we do expect at least ~1/32 inserts or deletes to cause a chunk boundary change.

Reply:
oof